### PR TITLE
Optimize Elkan K-means with OpenMP

### PR DESCRIPTION
**Pull request by MarkFischinger**
I've now applied OpenMP to the Elkan K-means implementation. Here are the results.

I ran a benchmark with the following settings:
```cpp
Number of data points: 100,000
Number of dimensions: 50
Number of clusters: 10
Maximum iterations: 100
```

**Here's what I found:**
```
Original version: 9.42804 seconds
OpenMP version: 7.19841 seconds
```
This represents a **23.6% decrease** in execution time.

### CODE REVIEW

**Comment by shrit** [comment]
@MarkFischinger looks good to me, I will approve it after the CI  checks out. In the meantime, Would you please show the performance on another test with a higher order of data points? such as 10M with at least 500 dimensions?

**Comment by conradsnicta** [comment]
> This represents a **23.6% decrease** in execution time.

This seems a bit low given that the code is now parallelised.  A bog-standard kmeans algorithm should be around twice as fast on a 2 core machine.  The Elkan approach is clearly not the same, but I'd expect there'd be more of a speedup.

Image extracted from https://doi.org/10.1109/ICSPCS.2017.8270510

8270510-fig-2-source-large

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:89]: Won't removing this break the centroid calculation?

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:157]: The critical sections are probably the pain points that prevent better speedup here.  I think that if we want to OpenMP-ize more than just distance calculations, we may need to hold sets of centroids for each individual core, and then merge them all together at the end.  (Writing that in a way that is agnostic to whether or not OpenMP is available may involve ifdefs... maybe even as simple as:

```
#ifdef _OPENMP // not sure this is the right macro, be sure to double-check
const size_t threads = omp_get_num_threads();
#else
const size_t threads = 1;
#endif
```

and then base the rest of your code around `threads` or something like this.  Just one of many possible ideas.)

**Comment by MarkFischinger** [comment]
> This seems a bit low given that the code is now parallelised [...] 

@conradsnicta I reran the benchmarks, but the speed remained the same. After making some additional adjustments (also mentioned by @rcurtin) to the elkan kmeans, here are the updated benchmark results:

**Settings:** 
```cpp
  const size_t numPoints = 100000;  // Number of data points.
  const size_t numDimensions = 50;  // Number of dimensions.
  const size_t numClusters = 10;    // Number of clusters.
  const size_t maxIterations = 100; // Maximum number of iterations.
```

**Benchmark:**
```cpp
  ~/gsoc/mlpack-benchmarks/src/kmeans$ ./elkan
  Running benchmark...
  Benchmark completed.
  Total time: 4.71125 seconds
```
So I got approximately the results we wanted and fixed some issues. Thank you for your feedback/comment :)

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:94]: When I take a look at the naive strategy, the idea there is to just create a local centroid object inside each thread. If you applied that strategy here too (and just used a `#pragma omp critical` when summing centroids) I think the code would be much simpler and you wouldn't need `numThreads`.  Let me know what you think.

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:193]: No need for the ifdef; the compiler will ignore the pragma if OpenMP is not enabled. :+1:

**Review by rcurtin — COMMENTED** [review]
Sorry this one has sat so long---my schedule is finally clearing up and I can get back to looking into these.  I think this one is pretty close and the speedup is nice!  I just have a few simple structural comments and similar; let me know if I can clarify anything. :+1:

**Comment by MarkFischinger** [comment]
@rcurtin @shrit I removed the OpenMP import and all the `MLPACK_USE_OPENMP` tags, fixed the line lengths and implemented ryan's idea that included local centroids for each thread.

Ryan's suggestion did yield a better execution time: 
```
Running benchmark for ElkanKMeans with 1000 points, 10 clusters, and 50 dimensions...
Total time: 0.04244 seconds
Running benchmark for ElkanKMeans with 5000 points, 10 clusters, and 50 dimensions...
Total time: 0.190035 seconds
Running benchmark for ElkanKMeans with 10000 points, 10 clusters, and 50 dimensions...
Total time: 0.353643 seconds
Running benchmark for ElkanKMeans with 50000 points, 10 clusters, and 50 dimensions...
Total time: 2.18762 seconds
Running benchmark for ElkanKMeans with 100000 points, 10 clusters, and 50 dimensions...
Total time: 4.59366 seconds
```

So the code before executed in this time:
 ```
Running benchmark for ElkanKMeans with 1000 points, 10 clusters, and 50 dimensions...
Total time: 0.0752948 seconds
Running benchmark for ElkanKMeans with 5000 points, 10 clusters, and 50 dimensions...
Total time: 0.356647 seconds
Running benchmark for ElkanKMeans with 10000 points, 10 clusters, and 50 dimensions...
Total time: 0.751239 seconds
Running benchmark for ElkanKMeans with 50000 points, 10 clusters, and 50 dimensions...
Total time: 4.68356 seconds
Running benchmark for ElkanKMeans with 100000 points, 10 clusters, and 50 dimensions...
Total time: 10.187 seconds
```

My variant before ryan's suggestion (called "optimized" in the graph below): 
```
Running benchmark for ElkanKMeans with 1000 points, 10 clusters, and 50 dimensions...
Total time: 0.055524 seconds
Running benchmark for ElkanKMeans with 5000 points, 10 clusters, and 50 dimensions...
Total time: 0.178682 seconds
Running benchmark for ElkanKMeans with 10000 points, 10 clusters, and 50 dimensions...
Total time: 1.70825 seconds
Running benchmark for ElkanKMeans with 50000 points, 10 clusters, and 50 dimensions...
Total time: 3.66519 seconds
Running benchmark for ElkanKMeans with 100000 points, 10 clusters, and 50 dimensions...
Total time: 4.70217 seconds
```

I used this [script](https://gist.github.com/MarkFischinger/17f708b5ff11dfcbe800bc1963452cce) for benchmarking.
Here is the data visualized:

benchmark_comparison

**Review by rcurtin — COMMENTED** [review]
This one looks pretty much ready to go to me, only a couple tiny comments.  I also mentioned the custom reduction idea that I mentioned in the Hamerly k-means PR again; we don't have to do that in this PR, but if you find the idea is worthwhile in the other PR and quick enough to do, might as well do it here too---let me know what you think.

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:84]:
```suggestion
      reduction(colAdd: counts) reduction(+: distanceCalculations)
```
Style fix (double indent on wrapped lines)

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:90]: We should keep this `continue`---otherwise the centroids will be computed twice for iterations where we take this `if`.

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:97]: No need to change this comment (I imagine it is because of the extra indentation that later got reverted with the custom OpenMP reductions); can you revert it to the original?

**Review by rcurtin — COMMENTED** [review]
The new custom reductions make this really nice and clean!  I think it will have to wait until #3761 gets merged (so that `omp_reductions.hpp` is available).  Just a couple small comments, then from my side things look good. :+1:

File: src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp
Comment (rcurtin) [inline src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp:155]: Oh!  I see now.  This has moved into the `else` condition, so whether or not there is a `continue` in the `if` body is now irrelevant.  So, up to you whether you want to keep it like this and remove the `continue`, or revert this.  (Personally I think it is cleaner like this but without the `continue`, but, up to you which to do. :+1:)

**Review by shrit — APPROVED** [review]

**Pull request merged**
